### PR TITLE
Fix problem where clone adds core.worktree due to path case differences

### PR DESCRIPTION
**Pull request by SyntevoAlex**
A fix for #2569

### CODE REVIEW

**Comment by dscho** [comment]
Hmm. This patch will retry under more circumstances than just when the file does not exist. I would have expected the patch to check whether the `GetFinalPathByHandle()` function failed because the item (it's not necessarily a file, right? Could be a directory or symlink, too) doesn't exist and only do the last component magic in that case.

Also, `[:upper:]`? I seem to remember that we use `A-Z` instead...

**Comment by SyntevoAlex** [comment]
> Hmm. This patch will retry under more circumstances than just when the file does not exist. I would have expected the patch to check whether the `GetFinalPathByHandle()` function failed because the item (it's not necessarily a file, right? Could be a directory or symlink, too) doesn't exist and only do the last component magic in that case.

Correct, it always retries. I could change it to only retry on `ERROR_FILE_NOT_FOUND`, but it seems that is going to be a harmful change. The only benefit I see is optimization, but this function only seems to be called very rarely and in the majority of cases it will not fail, so optimization value is near zero. At the same time, code will be more complicated (now the helper function will need to return 3 results), and I'm not quite sure if `ERROR_FILE_NOT_FOUND` is the only possible "interesting" error here. I will implement that if you still think that it should be done.

> Also, `[:upper:]`? I seem to remember that we use `A-Z` instead...

Thanks! I'm quite a newbie in shell scripts, so I just grabbed something from Stack Overflow and made sure it works. I wonder how `A-Z` is better, but nevermind, it indeed seems to be the agreed way in test scripts.

**Comment by SyntevoAlex** [comment]
Hmm, I didn't figure that other errors in last component could introduce difference between `mingw_strbuf_realpath()` and `strbuf_realpath()`.

I have tested 4 cases:
1. `C:\access_denied_folder\1.txt`
   `mingw_strbuf_realpath()` fails on path #1 with `ERROR_ACCESS_DENIED`.
   `strbuf_realpath()` fails with `ERROR_ACCESS_DENIED`.
2. `C:\access_denied_folder`
   `mingw_strbuf_realpath()` fails on path #1 with `ERROR_ACCESS_DENIED`.
   `mingw_strbuf_realpath()` succeeds on path #2.
   `strbuf_realpath()`, if forced to continue, succeeds.
3. `C:\broken_symlink\1.txt`
   `mingw_strbuf_realpath()` fails on path #1 with `ERROR_PATH_NOT_FOUND`.
   `mingw_strbuf_realpath()` fails on path #2 with `ERROR_FILE_NOT_FOUND`.
	`strbuf_realpath()` fails with `ERROR_PATH_NOT_FOUND`.
4. `C:\broken_symlink`
   `mingw_strbuf_realpath()` fails on path #1 with `ERROR_FILE_NOT_FOUND`.
   `mingw_strbuf_realpath()` succeeds on path #2.
   `strbuf_realpath()`, if forced to continue, succeeds.

Observations:
1. `ERROR_FILE_NOT_FOUND` is also returned for broken symlinks.
2. Current behavior of `mingw_strbuf_realpath()` matches that of `strbuf_realpath()`.
3. If only `ERROR_FILE_NOT_FOUND` is retried, behavior of `mingw_strbuf_realpath()` will not match `strbuf_realpath()`.

Conclusion:
I understand that it should continue to retry on all errors.

**Comment by dscho** [comment]
> I wonder how `A-Z` is better

You probably tested with a GNU version of `tr`, right? These tests need to run in all kinds of settings, also using older `tr`, and BSD versions of `tr`. In general, it is always a good idea to imitate existing code when patching a legacy code base. There is no `tr` call in Git that uses `[:upper:]`.

> Correct, it always retries.

That is bad. If the first call fails for unrelated reasons, we should return _that_ error, not overwrite it with a second call's error.

> I could change it to only retry on `ERROR_FILE_NOT_FOUND`, but it seems that is going to be a harmful change.

I fail to see how this would be harmful.



> At the same time, code will be more complicated (now the helper function will need to return 3 results)

Oh, I see. I did not actually mean to keep that helper. It is misnamed, and I think it is also slightly overkill ;-)

File: t/t5601-clone.sh [hunk 1]
```diff
@@ -68,6 +68,14 @@ test_expect_success 'clone respects GIT_WORK_TREE' '
 
 '
 
+test_expect_success MINGW,CASE_INSENSITIVE_FS 'core.worktree is not added due to path case' '
+
+	mkdir CAPITALCASE &&
+	lower_path=$( cygpath -w "$(pwd)/CAPITALCASE" | tr A-Z a-z ) &&
```
Comment (dscho) [inline t/t5601-clone.sh:74]: Please remove the spaces after the opening, and before the closing brace.

File: t/t5601-clone.sh [hunk 2]
```diff
@@ -68,6 +68,14 @@ test_expect_success 'clone respects GIT_WORK_TREE' '
 
 '
 
+test_expect_success MINGW,CASE_INSENSITIVE_FS 'core.worktree is not added due to path case' '
+
+	mkdir CAPITALCASE &&
```
Comment (dscho) [inline t/t5601-clone.sh:73]: Isn't it "UPPERCASE"?

File: t/t5601-clone.sh (hunk 1 above)
Comment (dscho) [reply t/t5601-clone.sh:74]: Oh, and why not simply using `$(pwd)/capitalcase`? It is sufficient for the case to differ only in part of the path, right?

File: t/t5601-clone.sh [hunk 3]
```diff
@@ -68,6 +68,14 @@ test_expect_success 'clone respects GIT_WORK_TREE' '
...
+	lower_path=$( cygpath -w "$(pwd)/CAPITALCASE" | tr A-Z a-z ) &&
+	git clone src "$lower_path" &&
+	test -z $(cd "CAPITALCASE" && git config core.worktree)
```
Comment (dscho) [inline t/t5601-clone.sh:76]: Better: `test -z "$(git -C CAPITALCASE config core.worktree)"`

File: t/t5601-clone.sh [hunk 4]
```diff
@@ -68,6 +68,14 @@ test_expect_success 'clone respects GIT_WORK_TREE' '
 
 '
 
+test_expect_success MINGW,CASE_INSENSITIVE_FS 'core.worktree is not added due to path case' '
+
+	mkdir CAPITALCASE &&
+	lower_path=$( cygpath -w "$(pwd)/CAPITALCASE" | tr A-Z a-z ) &&
+	git clone src "$lower_path" &&
```
Comment (dscho) [inline t/t5601-clone.sh:75]: It's actually the lower _case_ path. But I think `lower` would be plenty sufficient.

File: compat/mingw.c [hunk 5]
```diff
@@ -1287,35 +1287,61 @@ struct tm *localtime_r(const time_t *timep, struct tm *result)
...
-	if (xutftowcs_path(wpath, path) < 0)
-		return NULL;
+	if (xutftowcs_path_ex(wpath, path, MAX_PATH, path_len, MAX_PATH, 0) < 0)
+		return -1;
 
 	h = CreateFileW(wpath, 0,
 			FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
 			OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
 	if (h == INVALID_HANDLE_VALUE)
-		return NULL;
```
Comment (dscho) [inline compat/mingw.c:1304]: So is it this function call that is failing, or the `GetFinalPathNameByHandleW()` one?

File: t/t5601-clone.sh (hunk 1 above)
Comment (homo-programmatis) [reply t/t5601-clone.sh:74]: > Please remove the spaces after the opening, and before the closing brace.

I wanted to do that myself, but the rest of this file seems to use this style. If you still want it removed, I will do.

Comment (SyntevoAlex) [reply t/t5601-clone.sh]: > Please remove the spaces after the opening, and before the closing brace.

Copied that from line 170. This files seems to use both styles. Fixed as requested.

Comment (SyntevoAlex) [reply t/t5601-clone.sh]: > Oh, and why not simply using `$(pwd)/capitalcase`? It is sufficient for the case to differ only in part of the path, right?

Sorry, I don't quite understand. If you mean that `C:/` will be different anyway, I was worried that tests could be run from `\\SERVER\share`

File: t/t5601-clone.sh (hunk 2 above)
Comment (SyntevoAlex) [reply t/t5601-clone.sh]: You're right. Fixed. I was thinking "capital letter case".

File: t/t5601-clone.sh (hunk 3 above)
Comment (SyntevoAlex) [reply t/t5601-clone.sh]: Fixed. Copy&pasted that from t0001.

File: t/t5601-clone.sh (hunk 4 above)
Comment (SyntevoAlex) [reply t/t5601-clone.sh]: Changed to `lower` as requested.

File: compat/mingw.c (hunk 5 above)
Comment (SyntevoAlex) [reply compat/mingw.c:1304]: It's `CreateFile()` that fails.

Comment (dscho) [reply compat/mingw.c:1304]: Of course, because of the `OPEN_EXISTING` flag.

So here would be a splendid spot to insert `if (h == INVALID_HANDLE_VALUE && GetLastError()) ...`. The easiest way would actually to look for the last path separator in both `path` and `wpath`; That way, you do not have to duplicate the last component, but can trim the `wpath` and by storing its position within `path` in a new local variable (that can be initialized to ""), you can then append it after the `xwcstoutf()` call below (actually after verifying that that call worked successfully).

That's not a ton more code, actually. Essentially something like this:

```c
if (h == INVALID_HANDLE_VALUE && GetLastError() == FILE_NOT_FOUND) {
    /* remember the last component and trim `wpath` for now */
    wchar_t *p = wpath + wcslen(wpath);

    while (p != wpath)
        if (*(--p) == L'/' || *p == L'\\')
            break; /* found start of last component */
    if (p != wpath && (last_component = find_last_dir_sep(path)) {
        *p = L'\0';
        h = CreateFileW(wpath, 0,
			FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
			OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
    }
}
```

And then later

```c
	len = xwcstoutf(resolved->buf, normalize_ntpath(wpath), len);
	if (len < 0)
		return NULL;
	if (last_component)
		strbuf_addstr(resolved, last_component);
```

File: t/t5601-clone.sh (hunk 1 above)
Comment (dscho) [reply t/t5601-clone.sh]: No, I meant that the `uppercase` will be different from `UPPERCASE` anyway.

File: t/t5601-clone.sh [hunk 6]
```diff
@@ -68,6 +68,14 @@ test_expect_success 'clone respects GIT_WORK_TREE' '
 
 '
 
+test_expect_success MINGW,CASE_INSENSITIVE_FS 'core.worktree is not added due to path case' '
+
+	mkdir UPPERCASE &&
+	lower=$(cygpath -w "$(pwd)/UPPERCASE" | tr A-Z a-z) &&
+	git clone src "$lower" &&
+	test -z $(git -c UPPERCASE config core.worktree)
```
Comment (dscho) [inline t/t5601-clone.sh:76]: Oh, but that won't work without double quotes around the `$(...)`, will it?

File: compat/mingw.c [hunk 7]
```diff
@@ -1287,35 +1287,61 @@ struct tm *localtime_r(const time_t *timep, struct tm *result)
...
+
+	if (realpath_existing_file(resolved, path, -1) >= 0)
+		return resolved->buf;
+
+	// strbuf_realpath() allows last path component to not exist.
+	// Resolving full path failed, now it's time to try without last component.
 
+	// Retry on all errors to better match behavior of `strbuf_realpath()`.
+	// For example, for no-access last directory, `strbuf_realpath()` ignores
+	// that (because it doesn't try to `CreateFile()`) and still succeeds.
```
Comment (dscho) [inline compat/mingw.c:1332]: Good comment, but in Git's source code it cannot be C++ style, it has to be `/* ... */`.

File: compat/mingw.c [hunk 8]
```diff
@@ -1287,35 +1287,61 @@ struct tm *localtime_r(const time_t *timep, struct tm *result)
...
+	last_component = find_last_dir_sep(path);
+	if (!last_component)
+		return NULL;
+
+	if (realpath_existing_file(resolved, path, last_component - path) < 0)
+		return NULL;
+
+	// Force / instead of \ in appended component
+	strbuf_addch(resolved, '/');
+	strbuf_addstr(resolved, last_component + 1);
```
Comment (dscho) [inline compat/mingw.c:1343]: Oh, good point, this comes _after_ the `normalize_ntpath()` call.

File: t/t5601-clone.sh (hunk 6 above)
Comment (SyntevoAlex) [reply t/t5601-clone.sh]: Works for me, but again I'm such a shell script newbie. I added double quotes, it also seems to work fine. I also reworked this test to avoid the possibility of other errors in `git config` to be incorrectly recognized as missing key.

File: compat/mingw.c (hunk 7 above)
Comment (SyntevoAlex) [reply compat/mingw.c:1332]: Good catch!

File: t/t5601-clone.sh (hunk 1 above)
Comment (SyntevoAlex) [reply t/t5601-clone.sh]: Indeed, I was too fixed on the idea that `c:/` causes the problem and wanted to test both, but that's probably an overkill. I have changed as suggested now.

File: t/t5601-clone.sh [hunk 9]
```diff
@@ -68,6 +68,13 @@ test_expect_success 'clone respects GIT_WORK_TREE' '
 
 '
 
+test_expect_success MINGW,CASE_INSENSITIVE_FS 'core.worktree is not added due to path case' '
```
Comment (dscho) [inline t/t5601-clone.sh:71]: Now that you no longer use `cygpath`, how about dropping the `MINGW` prereq?

**Comment by dscho** [comment]
Awesome work, @SyntevoAlex, keep it up! I'm crashing right now, but will come back tomorrow.

File: t/t5601-clone.sh (hunk 1 above)
Comment (SyntevoAlex) [reply t/t5601-clone.sh]: I also dropped `cygpath -w` part, it seems that cygwin converts paths automatically when starting git.exe

File: t/t5601-clone.sh (hunk 9 above)
Comment (SyntevoAlex) [reply t/t5601-clone.sh:71]: Done.

**Comment by SyntevoAlex** [comment]
Thanks for a quick review!

On a different topic, this patch had me thinking: why don't you strike an agreement with Junio to just wildcard upstream everything that only affects MINGW? Or maybe with only a surface review?

File: compat/mingw.c (hunk 8 above)
Comment (SyntevoAlex) [reply compat/mingw.c:1345]: Right.

File: compat/mingw.c (hunk 5 above)
Comment (SyntevoAlex) [reply compat/mingw.c:1304]: It seems that the core question here is whether to narrow retrying to `FILE_NOT_FOUND` or not. Previously I already shown that with `FILE_NOT_FOUND` only, `mingw_strbuf_realpath()` may fail where `strbuf_realpath()` does not, which could again invite issues like the PR is trying to solve. Just to remind, `core.worktree` problem begins when `mingw_strbuf_realpath()` fails and `strbuf_realpath()` succeeds.

I would say that `mingw_strbuf_realpath()` shall never fail where `strbuf_realpath()` succeeds.

Now, what other errors could cause that? For instance, `xutftowcs_path_ex()` may fail, because the last component is too long. Again, `strbuf_realpath()` will probably succeed here, and my design will succeed, but your design will fail and fall back to `strbuf_realpath()`. It's a bit hard to really judge if this is a good or bad thing, but again, I would rather prefer to stay closer to `strbuf_realpath()`.

It's a bit hard to predict which design may have more unexpected problems in it. All of this comes from the difference between `lstat()` (used in original `strbuf_realpath()`) and `CreateFile()`, where the first is more forgiving - it can tolerate things like no-access folders and broken symlinks.

If we agree to retry on all errors, then I'm not sure if your design is better. It's roughly the same amount of code, but it's harder to follow, as it has more parallel things to keep in mind while reading. For example, simultaneously removing (in different ways) the last component from two strings instills doubt and takes an extra read to verify. In my design, I think it's easier to grasp the second function once the first one is understood. It it's not about code clarity, then the other possible reason is performance. Performance-wise, the only difference is that my design will execute an extra `xutftowcs_path_ex()`, and your design will instead search for last component in `wchar*`. That's probably a very small difference compared to cost of `CreateFile()`.

To summarize, if you're confident and want your way, say so and I will implement.

Comment (SyntevoAlex) [reply compat/mingw.c:1304]: A different observation: I'm suspecting that the root cause is even deeper. To my understanding, you implemented `mingw_strbuf_realpath()` to solve problem with junctions. During my debugging adventures, I noticed that `file_attr_to_st_mode()` does not recognize junctions as `S_IFLNK`. That definitely raised my eyebrow.

Could it be that entire `mingw_strbuf_realpath()` needs to be removed again, and `file_attr_to_st_mode()` fixed so that `strbuf_realpath()` works as intended? That will put an end to all cases where `mingw_strbuf_realpath()` and `strbuf_realpath()` disagree.

What do you think?

Comment (dscho) [reply compat/mingw.c:1304]: > During my debugging adventures, I noticed that `file_attr_to_st_mode()` does not recognize junctions as `S_IFLNK`.

Yes, that is intentional. see e.g. the discussion in https://github.com/git-for-windows/git/pull/437.

> It seems that the core question here is whether to narrow retrying to `FILE_NOT_FOUND` or not.

I guess so.

> Previously I already shown that with `FILE_NOT_FOUND` only, `mingw_strbuf_realpath()` may fail where `strbuf_realpath()` does not

And if `mingw_strbuf_realpath()` fails, that's okay, because it then returns `NULL` and `strbuf_realpath()` continues with the non-Windows-specific code. Remember, `mingw_strbuf_realpath()` is _not_ a replacement of `strbuf_realpath()`, it is a potential shortcut called by the latter.

> Just to remind, `core.worktree` problem begins when `mingw_strbuf_realpath()` fails and `strbuf_realpath()` succeeds.

Sure, but the changes I proposed fix that, too. And their scope is so narrow that I am relatively confident that no unintended side effects are introduced. Retrying on all errors, on the other hand, _does_ strike me as _asking_ for unintended side effects.

> I would say that `mingw_strbuf_realpath()` shall never fail where `strbuf_realpath()` succeeds.

That was not the intention when I introduced the former as a potential shortcut by calling it from the latter.

> To my understanding, you implemented `mingw_strbuf_realpath()` to solve problem with junctions.

Yes, but I _also_ used the opportunity to use native Win32 API to resolve symlinks rather than doing it (tediously) "by hand".

> It's roughly the same amount of code,

Not really:

```
 compat/mingw.c   | 42 +++++++++++++++++++++++++++++++++++-------
```

vs

```
 compat/mingw.c   | 32 ++++++++++++++++++++++++++++++++
```

> but it's harder to follow

I find it harder to follow the two-function version, actually, especially because the return value type was changed (which was unnecessary) and because it is _so_ unclear in the two-function version whether we are retrying for a legitimate reason or not.

> Afterall, I have a bad gut feeling about two different `strbuf_realpath()` implementations, which are prone to disagreeing because they use quite different APIs.

First of all, they are not _different_ implementations. One function calls the latter (a platform-specific shortcut, where available).

And the only reason why they use different APIs is because `strbuf_realpath()` can only rely on a subset of POSIX functionality, and therefore cannot use platform-specific functions. That's why `platform_strbuf_realpath` is a thing, to abstract that away from the generic implementation.

I do agree, however, that it is really tricky to navigate all of the issues with junction points vs symlinks. Which is the reason why I want to stay away from making sweeping changes as much as I can, and use more surgical ways to address the issue you found.

I pushed my proposed `fixup!` to my fork (under the same branch name you chose, to make it easier for you) and the test suite is currently running: https://github.com/dscho/git/runs/833879717?check_suite_focus=true.

Comment (SyntevoAlex) [reply compat/mingw.c:1304]: Thanks for taking time to prepare a fixup! It looks reasonable to me, and I squashed it as is.
